lib/main.js

```diff
@@ -0,0 +1,28 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+import bedrock from 'bedrock';
+const {config, util: {extend}} = bedrock;
+import fs from 'fs';
+import jsYaml from 'js-yaml';
+import logger from './logger.js';
+import path from 'path';
+
+import './config.js';
+
+const namespace = 'config-yaml';
+const cfg = config[namespace];
+
+bedrock.events.on('bedrock.start', () => {
```

Comment:
This bedrock event seems like it would be too late in the process (it is after bedrock.init and so on ... and modules would miss config changes via the yaml file). What about using the `bedrock.configure` event instead?

Reply:
The concern I have about `bedrock.configure` is that we don't want an ordering issue on an event.  We need this module to be LAST in any list of configuration things.  Which is why it made sense to me to put it on `start`.  

I just double checked bedrock-mongodb and discovered that my memory of when it's init took place was incorrect.  It iinitializes on `bedrock.init` as does bedrock-redis. Therefore, as you say, `bedrock.start` is too late.

https://github.com/digitalbazaar/bedrock-mongodb/blob/master/lib/index.js#L67
https://github.com/digitalbazaar/bedrock-redis/blob/master/lib/index.js#L26

I have been trying to test this code in development this week which would have revealed this bug, but I haven't gotten to it.

I had earlier contemplated the possibility that a new event may be required here.  

What is not an acceptable proposition in my view is to have any solution that requires that `bedrock-config-yaml` be required in some particular order.

@dlongley do you see a solution that allows us to work with the `bedrock.configure` event as it exists today?

Reply:
@mattcollier -- if we can't specify the order, no, I don't see a solution here. But I don't know why we can't specify the order. Maybe I don't understand how this would be used in practice. I was expecting it to be something like this:

// the file that runs the thing
```js
// require all the bedrock modules you need
require('bedrock');
// require('bedrock-*');

// require bedrock-config-yaml last
require('bedrock-config-yaml');

// start bedrock
bedrock.start();
```

Is this not accurate? If it is, then it seems that there should be ordering issue -- users do need to know to place the module last, but that is already common bedrock practice -- include your (peer-)dependencies first... and `bedrock-config-yaml` will be making changes to its (peer-)dependents' configs.

Reply:
To add -- there's no end to the "just add another event to bedrock that runs at the end" problem. That's where `require` order comes into play. If `bedrock-config-yaml` is modifying bedrock config, then it should run those changes on `bedrock.configure`. What happens if we add `bedrock.configure-done-i-swear` and then we have TWO modules that need to "run at the end"? This is only ever solved by having your application require the modules in the expected order.

Reply:
@dlongley you have described the issue precisely.  However, I disagree with your conclusion.  I'm looking for a solution that does not require things to be required in a particular order.  This is a solved problem in the rest of our stack through the use of peer dependencies.  If `bedrock-foo` wants to change the config of `bedrock-bar` it must explicitly specify a peer dependency and require it before it attempts to configure it.  Obviously having bedrock-config-yaml have a peer dep and require all the things is not an option.

I have had to debug enough issues related to the solution you proposed that I am convinced that we must have a better solution.

Reply:
@mattcollier, I'm not suggesting that `bedrock-config-yaml` have peer deps for all things, rather, I'm suggesting that whomever creates and uses `bedrock-config-yaml` will be using configuration variables from its own peer deps. It should include those in order, with `bedrock-config-yaml` alst.

Reply:
If, even after this is clear, you still believe we need another solution, I'm all ears, but I don't think teaching devs that you need to include `bedrock-config-yaml` last is too terrible. Here's a possible way to enforce it:

We could have `bedrock-config-yaml` listen on `bedrock.configure` and apply its config changes then. It could also check to see if there are any other listeners that come after it on that event, and if so, throw an error. This error could be disabled via config but would be thrown by default. Thoughts? That would force devs to ensure they put it last.

Reply:
In `bedrock-config-yaml`:

// in its config.js
```js
const cfg = bedrock.config['config-yaml'] = {};
// bikeshed
cfg.mustBeLastBedrockImport = true;
```

```js
bedrock.events.on('bedrock.configure', async applyConfig() => {
  const cfg = bedrock.config['config-yaml'];
  if(cfg.mustBeLastBedrockImport) {
    const listeners = bedrock.events.listeners('bedrock.configure');
    if(listeners[listeners.length - 1] !== applyConfig) {
      throw new Error('"bedrock-config-yaml" must be the last import.');
    }
  }
  // apply config
});
```

Reply:
@dlongley that's a pretty good idea.  It gave me the idea that maybe this is a core bedrock feature and therefore it is effectively hard coded after the `bedrock.configure` event.  This would sidestep all these considerations.  

I think there is a strong argument to be made for why this configuration feature is a candidate for core.

Reply:
I'm not sure how this would look as a core feature -- so I'm not quite sure what you're proposing (and I'm not yet for or against so I don't understand). I wouldn't be for making `bedrock-config-yaml` a special call out in bedrock, for example. If you're saying we would put into core something that would allow some kind of generic "check if I'm last" thing that a module could ask, that seems helpful. Any feature added to bedrock core needs to be very generic or we end up having to upgrade the whole ecosystem when it requires changes/maintenance, etc. (as you know).

Reply:
Another option is to only add a bedrock event listener once bedrock has started event processing... at either:

`bedrock-cli.init` or `bedrock-cli.ready`

Then, it would run last regardless of import order -- unless some other module is doing the same, but those modules are both trying to accomplish the goal of running last, so you'd have to ensure you order those particular two in the order you want (very special case).

Reply:
> Another option is to only add a bedrock event listener once bedrock has started event processing

This sort of thing sounds like it would be very difficult to debug.

What I was suggesting earlier as to how this would be implemented as a core feature is that it would not be hung on an event listener at all.  This code:

https://github.com/digitalbazaar/bedrock-config-yaml/blob/e16b5125f30bcab6bb3a97a17b3b4b83c71a7fc6/lib/main.js#L18-L27

Would be hard coded after L492 here:

https://github.com/digitalbazaar/bedrock/blob/53bc54b4f59f50946fa5ec96680505866381990b/lib/bedrock.js#L491-L499

Which is to say we will always check for the existence of a YAML file.  This check could be optional based on the regular bedrock config.  The optionality in this implementation was provide by whether or not one had the module installed/required.

So config in bedrock core could be something like:
```
bedrock.config.yaml = {
  loadOnStart: false,
  path: '/etc/bedrock',
  file: 'config.yaml',
}
```

Reply:
@dlongley implemented the event handler check here, I think this is a good solution to address the usability concerns I expressed.

https://github.com/digitalbazaar/bedrock-config-yaml/commit/d3d618e1ccc4ebfbc435f2bfadd60211f496f7a6